index.md

```diff
@@ -1,16 +1,128 @@
+
+That sets up the third point: what will be necessary in order to turn the installed payload into a Jupyterlab-runnable kernel and a terminal environment with the payload contents configured and available?
+
+## Payload configuration
+
+Configuration of the payload environment has at least three components.
+
+Setting up the Jupyter kernel is the most straightforward.  If the payload can be represented in a Python virtual environment, `python -m ipykernel install` is all that needs to be done.  If the payload is more complicated (as the DM pipelines stack is), then it turns out that the kernel can be a shell script that does necessary setup and execs a python process as its last action.
+
+The next is setting up a terminal environment.  One way to do this is to set a sentinel environment variable at Jupyterlab launch, and then when the user requests a shell, in the user's `.bashrc` or moral equivalent, test that variable, and if it is set (indicating the user is running inside Jupyterlab), perform the necessary setup (environment activation, running an appropriate setup script, whatever) before yielding control to the user for interactive shell use.
+
+The third is the most difficult.  In Nublado, we have a spawner form.  That form contains a selection of prepulled images, a drop-down box to select an uncached image, a selector to choose container size, and two checkboxes useful for debugging and unwedging stuck environments.  I am not actually sure how generic that is, but I think erring on the side of "everyone will want something like this" is better than insisting it's DM-pipelines-specific.
+
+The purpose of the form in the RSP environment is basically twofold: first, to let the user select how much they need in the way of resources, and second to reflect our release cadence.  We have, effectively, "daily", "weekly", and "release" images and would like some way to present those to the user and update them without human intervention as new images are built.
+
+By having rules about how many of each we use, we can also decide which images to prepull.  Given our use cases, it is vital to prepull images, because they are so large that waiting for a spawn that needs to pull an image from a remote repository becomes quite an unpleasant experience.  If the images are cached to each node (which Nublado does), then for images displayed on the menu (rather than in the drop-down list of all images), the spawn is generally a matter of seconds rather than minutes.  We simply poll the image repository and pull the appropriately-tagged images periodically.  Most of the time, neither the tags nor the content at each tag has changed, and in this case, all the "pull" is the server replying that the image associated with a tag has a particular sha256 checksum, and the client saying "oh, I've got that one, nothing to do, then"; thus while we do many thousands of pulls a day, only a handful actually do much work.
```

Comment:
This is not really that unpleasant on Google Cloud due to the fast-start capabilities of Google's internal container registry, right? I thought we tested and it brought spawn times down to a minute or so for uncached images pulled from Google.

Minor correction: this is not how image prepulling works. It asks the nodes whether the images are already present and if so it does nothing. It doesn't pull all appropriately-tagged images periodically. We therefore do not do thousands of pulls a day; normally we only do pulls after a new daily or weekly is uploaded.

Reply:
Oh, that's right, and that's why we need read:tags and thus authentication now.

It's still pretty unpleasant.  It's about 3 minutes for "planb" containers (which are about 1.5 times the size of single-stack containers).